File: python/socketd/socketd/transport/ChannelAssistant.py
from typing import Any
from abc import ABC, abstractmethod
from socketd.core.module.Frame import Frame


class ChannelAssistant(ABC):
    @abstractmethod
    async def write(self, target: Any, frame: Frame) -> None:
        pass

    @abstractmethod
    def is_valid(self, target: Any) -> bool:
        pass

    @abstractmethod
    async def close(self, target: Any) -> None:
        pass

    @abstractmethod
    def get_remote_address(self, target: Any) -> str:
        pass

    @abstractmethod
    def get_local_address(self, target: Any) -> str:
        pass

File: go/socketd-example/channel_assistant.go
package main

import (
	"net"

	"socketd/transport/core"
	"socketd/transport/core/message"
)

var _ core.ChannelAssistant[*net.TCPConn] = new(TcpChannelAssistant)

type TcpChannelAssistant struct {
	Config core.Config
}

func NewTcpChannelAssistant(cfg core.Config) *TcpChannelAssistant {
	return &TcpChannelAssistant{
		Config: cfg,
	}
}

func (t *TcpChannelAssistant) GetLocalAddress(conn *net.TCPConn) net.Addr {
	return conn.LocalAddr()
}

func (t *TcpChannelAssistant) GetRemoteAddress(conn *net.TCPConn) net.Addr {
	return conn.RemoteAddr()
}

func (t *TcpChannelAssistant) IsValid(conn *net.TCPConn) bool {
	return true
}

func (t *TcpChannelAssistant) Write(conn *net.TCPConn, frame *message.Frame) (err error) {
	_, err = conn.Write(t.Config.GetCodec().Encode(frame))
	return err
}

func (t *TcpChannelAssistant) Close(conn *net.TCPConn) error {
	return conn.Close()
}
